Sources/Datadog/Core/FeaturesConfiguration.swift

```diff
@@ -0,0 +1,162 @@
+        let uploadURLWithClientToken: URL
+        let applicationID: String
+        let sessionSamplingRate: Float
+    }
+
+    /// Configuration common to all features.
+    let common: Common
+    /// Logging feature configuration or `nil` if the feature is disabled.
+    let logging: Logging?
+    /// Tracing feature configuration or `nil` if the feature is disabled.
+    let tracing: Tracing?
+    /// RUM feature configuration or `nil` if the feature is disabled.
+    let rum: RUM?
+}
+
+extension FeaturesConfiguration {
+    /// Builds and validates the configuration for all features.
+    ///
+    /// It takes two types received from the user: `Datadog.Configuration` and `AppContext` and blends them together
+    /// with resolving defaults and ensuring the configuration consistency.
+    ///
+    /// Throws an error on invalid user input, i.e. broken custom URL.
+    /// Prints a warning if configuration is inconsistent, i.e. RUM is enabled, but RUM Application ID was not specified.
+    init(configuration: Datadog.Configuration, appContext: AppContext) throws {
+        var logging: Logging?
+        var tracing: Tracing?
+        var rum: RUM?
+
+        let common = Common(
+            applicationName: appContext.bundleName ?? appContext.bundleType.rawValue,
+            applicationVersion: appContext.bundleVersion ?? "0.0.0",
+            applicationBundleIdentifier: appContext.bundleIdentifier ?? "unknown",
+            serviceName: configuration.serviceName ?? appContext.bundleIdentifier ?? "ios",
+            environment: try ifValid(environment: configuration.environment),
+            performance: .best(for: appContext.bundleType)
+        )
+
+        if configuration.loggingEnabled {
+            logging = Logging(
+                common: common,
+                uploadURLWithClientToken: try ifValid(
+                    endpointURLString: configuration.logsEndpoint.url,
+                    clientToken: configuration.clientToken
+                )
+            )
+        }
+
+        if configuration.tracingEnabled {
+            var autoInstrumentation: Tracing.AutoInstrumentation?
+
+            if !configuration.tracedHosts.isEmpty {
+                autoInstrumentation = Tracing.AutoInstrumentation(
+                    tracedHosts: configuration.tracedHosts,
+                    excludedHosts: [
+                        configuration.logsEndpoint.url,
+                        configuration.tracesEndpoint.url,
+                        configuration.rumEndpoint.url
+                    ]
+                )
+            }
+
+            tracing = Tracing(
+                common: common,
+                uploadURLWithClientToken: try ifValid(
+                    endpointURLString: configuration.tracesEndpoint.url,
+                    clientToken: configuration.clientToken
+                ),
+                autoInstrumentation: autoInstrumentation
+            )
+        }
+
+        if configuration.rumEnabled {
```

Comment:
this is what we actually have in Android without the error part. I am going to add a ticket on our end to send warning logs to the console in case this happens as I find it very useful.